src-control/Control/RefCount.hs

```diff
@@ -41,94 +95,39 @@ instance NFData (RefCounter m) where
-addReference RefCounter{countVar} = do
+incrementRefCounter :: PrimMonad m => RefCounter m -> m ()
+incrementRefCounter RefCounter{countVar} = do
     prevCount <- fetchAddInt countVar 1
-    assertWithCallStack (prevCount > 0) $ pure ()
+    assert (prevCount > 0) $ pure ()
 
-{-# SPECIALISE removeReference :: HasCallStack => RefCounter IO -> IO () #-}
+{-# SPECIALISE decrementRefCounter :: RefCounter IO -> IO () #-}
 -- | Decrease the reference counter by one.
 --
 -- The count must be known (from context) to be non-zero. Typically this will
 -- be because the caller has a reference already (that they took out themselves
 -- or were given).
-removeReference :: (HasCallStack, PrimMonad m, MonadMask m) => RefCounter m -> m ()
-removeReference RefCounter{countVar, finaliser} = mask_ $ do
-    prevCount <- fetchSubInt countVar 1
-    assertWithCallStack (prevCount > 0) $ pure ()
-    when (prevCount == 1) $ sequence_ finaliser
-
--- TODO: remove uses of this API. Eventually all references should be singular,
--- and not use patterns where if A contains B then N references on A becomes N
--- references on B. Instead this should be a single reference from A to B,
--- irrespective of the number of references to A.
-{-# SPECIALISE removeReferenceN :: HasCallStack => RefCounter IO -> Word64 -> IO () #-}
--- | Decrease the reference counter by @n@. @n@ must be a positive number.
---
--- The count must be known (from context) to be non-zero and at least as large
--- as @n@. Typically this will be because the caller has @n@ references already
--- (that they took out themselves or were given).
-removeReferenceN :: (HasCallStack, PrimMonad m, MonadMask m) => RefCounter m -> Word64 -> m ()
-removeReferenceN RefCounter{countVar, finaliser} n = mask_ $ do
-    -- n should be positive
-    assert (n > 0) $ pure ()
-    let !n' = fromIntegralChecked n
-    prevCount <- fetchSubInt countVar n'
-    -- the reference count must not already be 0, because then the finaliser
-    -- will have run already
-    assertWithCallStack (prevCount > 0) $ pure ()
-    -- the reference count can not go below zero
-    assertWithCallStack (prevCount >= n') $ pure ()
-    when (prevCount <= n') $ sequence_ finaliser
-
--- TODO: remove when removeReferenceN is removed
-{-# INLINABLE fromIntegralChecked #-}
--- | Like 'fromIntegral', but throws an error when @(x :: a) /= fromIntegral
--- (fromIntegral x :: b)@.
-fromIntegralChecked :: (HasCallStack, Integral a, Integral b, Show a) => a -> b
-fromIntegralChecked x
-  | x'' == x
-  = x'
-  | otherwise
-  = error $ printf "fromIntegralChecked: conversion failed, %s /= %s" (show x) (show x'')
-  where
-    x' = fromIntegral x
-    x'' = fromIntegral x'
+decrementRefCounter :: (PrimMonad m, MonadMask m) => RefCounter m -> m ()
+decrementRefCounter RefCounter{countVar, finaliser} =
+    mask_ $ do
+      prevCount <- fetchSubInt countVar 1
+      assert (prevCount > 0) $ pure ()
+      when (prevCount == 1) finaliser
```

Comment:
Yeah, because we don't want to decrement and then get interrupted before we run the finaliser. It's a good point that this should be mentioned in the API docs (probably for newRef, where the finaliser is given)

Reply:
Hmm, so in a lot of cases the `TempRegistry` will actually run the surrounding code already inside a `mask`. I've taken a note that I should check for such cases of double masks, because I'm looking at this anyway in the fault injection milestone

Reply:
Yes, that's a good point. Perhaps we should have a debug-only assert that we are already masked?

Reply:
Actually it's quite inconvenient to assert that, since it requires an extra `MonadMaskingState` constraint.

Reply:
Then maybe add a TODO, that we *could* reduce the number of masks

Reply:
Done.